File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/309-best-time-to-buy-and-sell-stock-with-cooldown-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function maxProfit(prices: number[]): number {
    const n = prices.length;
    const memo: Map<string, number> = new Map();
    function f(i: number, holding: boolean, cooldown: boolean): number {
        if (i === n) return 0;
        const key = `${i},${holding},${cooldown}`;
        if (memo.has(key)) return memo.get(key)!;
        let best = f(i + 1, holding, false);
        if (holding) {
            best = Math.max(best, prices[i] + f(i + 1, false, true));
        } else if (!cooldown) {
            best = Math.max(best, -prices[i] + f(i + 1, true, false));
        }
        memo.set(key, best);
        return best;
    }
    return f(0, false, false);
}

assert(maxProfit([1, 2, 3, 0, 2]) === 3);
assert(maxProfit([1]) === 0);
assert(maxProfit([]) === 0);
assert(maxProfit([5, 4, 3, 2, 1]) === 0);
assert(maxProfit([1, 2, 3, 4, 5]) === 4);
assert(maxProfit([1, 2]) === 1);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/309-best-time-to-buy-and-sell-stock-with-cooldown-approach2.py
from functools import lru_cache

def max_profit(prices: list[int]) -> int:
    n = len(prices)
    @lru_cache(maxsize=None)
    def f(i, holding, cooldown):
        if i == n:
            return 0
        best = f(i + 1, holding, False)
        if holding:
            best = max(best, prices[i] + f(i + 1, False, True))
        elif not cooldown:
            best = max(best, -prices[i] + f(i + 1, True, False))
        return best
    return f(0, False, False)

assert max_profit([1, 2, 3, 0, 2]) == 3
assert max_profit([1]) == 0
assert max_profit([]) == 0
assert max_profit([5, 4, 3, 2, 1]) == 0
assert max_profit([1, 2, 3, 4, 5]) == 4
assert max_profit([1, 2]) == 1
print("all tests pass")
